### PR TITLE
https: use CURL instead of raw hand made http

### PR DESCRIPTION
**Pull request by EXtremeExploit**
The only maybe real difference with the old implementation its that the MD5 checksum is not checked anymore (not like it was really necessary anyways) and you cant cancel/pause the download, again, not like it matters since the download is at most 40MB and it takes like 5 seconds to download, and besides you already get a message asking if you are sure you want to update

### CODE REVIEW

**Comment by Zangetsu38** [comment]
I don't wsant this code if you no can pause/resume...
i have set it no only for vita3k update

**Comment by Macdu** [comment]
It is possible to use curl_easy_pause  to pause a download using curl. Moreover, we can assume without a md5 check that the content downloaded from curl has not been corrupted.

**Comment by Zangetsu38** [comment]
All can get corrupted, curl or not, i can get corrupted file in Firefox so curl lol
Is stupid adding new dependency when is not needed, my code is good already for what i done

File: vita3k/gui/src/vita3k_update.cpp [hunk 1]
```diff
@@ -397,34 +353,9 @@ void draw_vita3k_update(GuiState &gui, EmuEnvState &emuenv) {
             if (state == DOWNLOAD)
                 download_update(emuenv.base_path.string());
         }
-        if (state == DOWNLOAD)
-            ImGui::EndDisabled();
     }
 
-    // Draw Cancel popup
```
Comment (Zangetsu38) [inline vita3k/gui/src/vita3k_update.cpp:404]: I don't want to add a module like curl which is totally useless here, and which also removes features that I had added, and which worked very well before, that doesn't make sense.

**Review by Zangetsu38 — CHANGES_REQUESTED** [review]

File: vita3k/https/src/https.cpp [hunk 2]
```diff
@@ -236,285 +65,77 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
-        if (accum_bits >= 8) {
-            accum_bits -= 8;
-            md5_bytes[i++] = static_cast<unsigned char>((accum >> accum_bits) & 0xFF);
-        }
-    }
+typedef int (*CurlDownloadCallback)(void *bar, double t, double d);
 
-    return convert_md5_bytes_to_str(md5_bytes);
-}
+CURL *curl_download;
```
Comment (Zangetsu38) [inline vita3k/https/src/https.cpp:93]: this is wrong, here you set general value for only one download in same time
my current code of https allow multi download in same time

**Comment by Zangetsu38** [comment]
after test yyour branch is vvery bad for now and broken totaly build

for resume:
so generate project take + 3 min compare before
it no found and build curl again anytime run script
and compile project no found with no found lib of curl 👌

**Comment by EXtremeExploit** [comment]
Should be ready now, im not implementing MD5 checksum check as ITS OBSOLETE (https://www.rfc-editor.org/rfc/rfc9110.html)
> IANA has updated the "Content-MD5" entry in the new registry to have a status of 'obsoleted' with references to [Section 14.15](https://www.rfc-editor.org/rfc/rfc2616#section-14.15) of [[RFC2616](https://www.rfc-editor.org/rfc/rfc9110.html#RFC2616)] (for the definition of the header field) and [Appendix B](https://www.rfc-editor.org/rfc/rfc7231#appendix-B) of [[RFC7231](https://www.rfc-editor.org/rfc/rfc9110.html#RFC7231)] (which removed the field definition from the updated specification).
Dont cry about MD5. If your file gets corrupted bet better internet or cope

File: vita3k/https/src/https.cpp [hunk 3]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
+            return 1; // Returning anything thats not 0 aborts the request
         }
+        return 0;
+    };
 
-        // Call progress callback function to update progress info and state
-        if (progress_callback)
-            progress_state = progress_callback(progress_percent, remaining_time);
+    auto start_time = get_current_time_ms();
+    auto callbackData = std::pair<uint64_t, ProgressCallback>(start_time, progress_callback);
```
Comment (bookmist) [inline vita3k/https/src/https.cpp:126]: I think a typedef for curl callback data type (`std::pair<uint64_t, ProgressCallback>`) should be used.

File: vita3k/https/src/https.cpp [hunk 4]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
+
+    curl_easy_setopt(curl_download, CURLOPT_URL, url.c_str());
+    curl_easy_setopt(curl_download, CURLOPT_WRITEFUNCTION, write_data);
+    curl_easy_setopt(curl_download, CURLOPT_NOPROGRESS, false); // Enable progress function
+    curl_easy_setopt(curl_download, CURLOPT_XFERINFODATA, &callbackData);
+    curl_easy_setopt(curl_download, CURLOPT_XFERINFOFUNCTION, static_cast<CurlDownloadCallback>(curl_callback));
```
Comment (bookmist) [inline vita3k/https/src/https.cpp:132]: Why static_cast needed?

Comment (EXtremeExploit) [reply vita3k/https/src/https.cpp]: my bad, fixed in last push

File: vita3k/https/src/https.cpp (hunk 3 above)
Comment (EXtremeExploit) [reply vita3k/https/src/https.cpp]: done

File: vita3k/https/src/https.cpp [hunk 5]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
-        else
-            LOG_WARN("Canceled by user, dowloaded size: {}/{}", downloaded_file_size, file_size);
-        return false;
-    }
+    fclose(fp);
+    curl_easy_cleanup(curl_download);
+    progress_callback(0, 0);
 
-    // Check if the downloaded file is corrupted
-    const auto downloaded_file_md5 = calculate_md5_file(output_file_path);
```
Comment (Zangetsu38) [inline vita3k/https/src/https.cpp:510]: Don't remove my code of check and verify md5
I have take time for coding it...
I see in curl you can get header, using it

File: vita3k/gui/src/vita3k_update.cpp [hunk 6]
```diff
@@ -175,41 +167,19 @@ static void download_update(const std::string &base_path) {
         const std::string archive_ext = ".zip";
 #endif
 
-        const auto latest_ver_path = base_path + "latest-ver";
         const auto vita3k_latest_path = base_path + "vita3k-latest" + archive_ext;
 
         const std::string version = std::to_string(git_version);
 
-        // check if vita3k_latest_path exist
```
Comment (Zangetsu38) [inline vita3k/gui/src/vita3k_update.cpp:183]: Keep it here

**Comment by Zangetsu38** [comment]
> Should be ready now, im not implementing MD5 checksum check as ITS OBSOLETE (https://www.rfc-editor.org/rfc/rfc9110.html)
> 
> > IANA has updated the "Content-MD5" entry in the new registry to have a status of 'obsoleted' with references to [Section 14.15](https://www.rfc-editor.org/rfc/rfc2616#section-14.15) of [[RFC2616](https://www.rfc-editor.org/rfc/rfc9110.html#RFC2616)] (for the definition of the header field) and [Appendix B](https://www.rfc-editor.org/rfc/rfc7231#appendix-B) of [[RFC7231](https://www.rfc-editor.org/rfc/rfc9110.html#RFC7231)] (which removed the field definition from the updated specification).
> 
> Dont cry about MD5. If your file gets corrupted get better internet or cope

Get better internet ? 🤦‍♂️ Just no remove it that all

File: vita3k/https/src/https.cpp [hunk 7]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
 
-    // Get the file size from the response header (Content-Length)
-    const auto file_size = get_file_size(response);
-    if (file_size == 0) {
-        LOG_ERROR("Failed to get file size");
-        return false;
-    }
+        // Calculate remaining time in seconds
+        const auto remaining_bytes = static_cast<double>(t - d);
+        const auto remaining_time = static_cast<uint64_t>((remaining_bytes / d) * elapsed_time_ms) / 1000;
```
Comment (Zangetsu38) [inline vita3k/https/src/https.cpp:107]: Why remove my code for get raimaining time ? 
Your code is bad, you need using double obligatory
And no remove my code inside pause state... 
I have no coded all for no reason, all is here for get it good works, here you just broken it

File: vita3k/https/src/https.cpp [hunk 8]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
-    }
+        ProgressState *callback_result = data->second(progress_percent, remaining_time);
 
-    // Init SSL and socket connection with the downloaded file size (if exists)
-    auto https = init(url, "GET", downloaded_file_size);
-    if (!https.ssl)
-        return false;
-
-    // Calculate the header size based on the HEAD of response
-    auto header_size = response.length();
```
Comment (Zangetsu38) [inline vita3k/https/src/https.cpp:406]: Again, no remove my code for pause and resume

File: vita3k/https/src/https.cpp [hunk 9]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
-    const auto file_size = get_file_size(response);
-    if (file_size == 0) {
-        LOG_ERROR("Failed to get file size");
-        return false;
-    }
+        // Calculate remaining time in seconds
+        const auto remaining_bytes = static_cast<double>(t - d);
+        const auto remaining_time = static_cast<uint64_t>((remaining_bytes / d) * elapsed_time_ms) / 1000;
 
-    // Get the downloaded file size
```
Comment (Zangetsu38) [inline vita3k/https/src/https.cpp:391]: I have said what ? No broken my pause/resume
Here you juste broken my all stuff have take crazy time’for nothing 🤦‍♂️

File: vita3k/https/src/https.cpp [hunk 10]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
+            // Check if the download has been aborted
+            if (!callback_result->download) {
+                return 1; // Returning anything thats not 0 aborts the request
             }
-        } else {
-            // Sleep for 100ms to not consume CPU
-            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        };
 
-            // Reset initial downloaded file size and start time to calculate remaining time correctly when resume download
```
Comment (Zangetsu38) [inline vita3k/https/src/https.cpp:479]: Your code no works here, you need keep my code when is on pause for update time and init file size
If not time raimaining is juste broken

File: external/CMakeLists.txt [hunk 11]
```diff
@@ -262,6 +262,23 @@ if(APPLE)
...
+
+find_package(CURL QUIET)
+if(NOT CURL_FOUND)
+	message("System curl not found, compiling curl from source")
+	include(FetchContent)
+	FetchContent_Declare(curl
+		URL https://github.com/curl/curl/releases/download/curl-8_1_2/curl-8.1.2.tar.xz
+	)
+	FetchContent_MakeAvailable(curl)
+	target_link_libraries(libcurl INTERFACE curl)
```
Comment (Macdu) [inline external/CMakeLists.txt:275]:
```suggestion
	message("System curl not found, compiling curl from source")
	option(HTTP_ONLY "disables all protocols except HTTP (This overrides all CURL_DISABLE_* options)" ON)
	include(FetchContent)
	FetchContent_Declare(curl
		URL https://github.com/curl/curl/releases/download/curl-8_1_2/curl-8.1.2.tar.xz
	)
	FetchContent_MakeAvailable(curl)
```

We got no need for other protocols (IMAP, LDAP...)
Also the curl CMake defines libcurl and not curl.

Comment (Macdu) [reply external/CMakeLists.txt:275]: Also put the curl code after the openssl one in this file.

File: vita3k/https/CMakeLists.txt [hunk 12]
```diff
@@ -6,5 +6,4 @@ add_library(
 )
 
 target_include_directories(https PUBLIC include)
-target_include_directories(https PRIVATE ${OPENSSL_INCLUDE_DIR})
-target_link_libraries(https PUBLIC util)
+target_link_libraries(https PUBLIC util libcurl)
```
Comment (Macdu) [inline vita3k/https/CMakeLists.txt:9]: target_include_directories(https PRIVATE libcurl)
target_link_libraries(https PUBLIC util)

File: external/CMakeLists.txt (hunk 11 above)
Comment (Macdu) [reply external/CMakeLists.txt:275]: Also it's better to use the git module as it won't update the file timestamps every time cmake is run.

File: vita3k/https/src/https.cpp [hunk 13]
```diff
@@ -236,285 +63,94 @@ std::string get_web_regex_result(const std::string &url, const std::regex &regex
...
-    if (downloaded_file_size < (file_size * 0.99)) {
-        if (progress_state.download)
-            LOG_ERROR("Downloaded size is not equal to file size, downloaded size: {}/{}", downloaded_file_size, file_size);
-        else
-            LOG_WARN("Canceled by user, dowloaded size: {}/{}", downloaded_file_size, file_size);
-        return false;
-    }
+    fclose(fp);
+    curl_easy_cleanup(curl_download);
+    progress_callback(0, 0);
```
Comment (Macdu) [inline vita3k/https/src/https.cpp:146]:
```suggestion
   if (progress_callback)
        progress_state = progress_callback(0, 0);
```

File: vita3k/util/include/util/net_utils.h [hunk 14]
```diff
@@ -1,8 +1,12 @@
 #pragma once
 
 #include <boost/algorithm/string/compare.hpp>
+#include <curl/curl.h>
```
Comment (Macdu) [inline vita3k/util/include/util/net_utils.h:4]:
```suggestion
```

curl is linked as a private library and it is not used in the header, do not include it here.

File: vita3k/util/src/net_utils.cpp [hunk 15]
```diff
@@ -327,4 +327,138 @@ bool socketSetBlocking(int sockfd, bool blocking) {
     return true;
 }
```
Comment (Macdu) [inline vita3k/util/src/net_utils.cpp:329]:
```suggestion
#include <curl/curl.h>
```

Add it here instead (at the top of the file)

Comment (EXtremeExploit) [reply vita3k/util/src/net_utils.cpp:331]: done

**Comment by Zangetsu38** [comment]
You still removed me from the dev team just for ignore my review on this pr...
Just before seeing that I was right and that no, it didn't compile, that there was indeed an error, and also that no, not everyone uses vcpkg...
But so compile should be fixed now...

But, For this PR, once again, for the moment the PR has very little interest.

First, it adds an optional dependency which increases the generation time of the initial project, and therefore compilation.

Secondly, it removes a functionality for the resume, which I implemented first on my vita3K update as a test for the sequel which I have already partly coded, I NEED it...

Thirdly, each time you launch the batch it wastes unnecessary time on curl and you end up having to recompile curl inside visual studio each time.

To solve all this it's simple, if curl is as practical and easy to use as you say, restore my code for the summary by adapting it for curl :)
if it's more complicated that literally means that curl has no interest here.
I remind you that it is only me who implemented and entirely codes the vita3K update and the https code.
So avoid destroying everything I do...
I have no problem with an improvement, but this is anything but the case... it is precisely a regression...
I didn't want to add a submodule so as not to make the project heavier for something that is actually optional (in the sense that the emulator works without it), so I made do with what I had at hand.

Move curl dependency on pre build version, this will resolve the issues I indicated just before, vcpkg also uses a pre-build version.
I tried to do it myself, but unfortunately, on the website the binary version does not contain a libcurl.lib, but only a libcurl.a
So the project compiles but it fails on the link

And also, reverse all language changes...
Just keep the resume feature, that's all...

**Review by Zangetsu38 — CHANGES_REQUESTED** [review]
Leave mu resume feature present...

File: vita3k/util/src/net_utils.cpp [hunk 16]
```diff
@@ -327,4 +332,137 @@ bool socketSetBlocking(int sockfd, bool blocking) {
...
+}
+
+typedef int (*CurlDownloadCallback)(void *c, long t, long d);
+
+bool download_file(std::string url, const std::string &output_file_path, ProgressCallback progress_callback) {
+    CURL *curl_download = curl_easy_init();
+    if (!curl_download)
+        return false;
+
+    CurlDownloadCallback curl_callback = +[](void *c, long t, long d) {
```
Comment (Macdu) [inline vita3k/util/src/net_utils.cpp:399]:
```suggestion
    CurlDownloadCallback curl_callback = +[](void *user_data, long bytes_downloaded, long bytes_total) {
```
Give more meaning to the variables

File: vita3k/util/src/net_utils.cpp [hunk 17]
```diff
@@ -327,4 +332,137 @@ bool socketSetBlocking(int sockfd, bool blocking) {
...
+        const auto remaining_time = static_cast<uint64_t>((remaining_bytes / d) * elapsed_time_ms) / 1000;
+
+        ProgressState *callback_result = data->second(progress_percent, remaining_time);
+
+        std::mutex mtx;
+        std::unique_lock<std::mutex> lock(mtx);
+
+        callback_result->cv.wait(lock, [&]() {
+            return !callback_result->pause;
+        });
```
Comment (Macdu) [inline vita3k/util/src/net_utils.cpp:422]:
```suggestion
        std::unique_lock<std::mutex> lock(callback_result->mutex);

        callback_result->cv.wait(lock, [&]() {
            return !callback_result->pause;
        });
```

File: vita3k/gui/src/vita3k_update.cpp [hunk 18]
```diff
@@ -415,15 +377,21 @@ void draw_vita3k_update(GuiState &gui, EmuEnvState &emuenv) {
         ImGui::SetCursorPos(ImVec2((POPUP_SIZE.x / 2.f) - LARGE_BUTTON_SIZE.x - (20.f * SCALE.x), POPUP_SIZE.y - LARGE_BUTTON_SIZE.y - (22.0f * SCALE.y)));
         if (ImGui::Button(common["no"].c_str(), LARGE_BUTTON_SIZE)) {
             progress_state.pause = false;
+            progress_state.cv.notify_one();
+
```
Comment (Macdu) [inline vita3k/gui/src/vita3k_update.cpp:381]:
```suggestion
            std::unique_lock<std::mutex> lock(progress_state.mutex);
            progress_state.pause = false;
            progress_state.cv.notify_one();

```

File: vita3k/util/include/util/net_utils.h [hunk 19]
```diff
@@ -31,6 +35,18 @@ struct parsedUrl {
     bool invalid = false;
 };
 
+struct ProgressState {
+    std::condition_variable cv;
```
Comment (Macdu) [inline vita3k/util/include/util/net_utils.h:39]:
```suggestion
    std::condition_variable cv;
    std::mutex mutex;
```

**Review by Macdu — APPROVED** [review]

**Comment by Zangetsu38** [comment]
I very like like you ignore my opinion on my proper stuff...

**Pull request merged**